ballerina/http_connection.bal

```diff
@@ -210,57 +213,92 @@ isolated function createStatusCodeResponse(StatusCodeResponse message, string? r
-    setPayload(message?.body, response, setETag);
-
+    string? mediaType = retrieveMediaType(message, returnMediaType);
+    setPayload(message?.body, response, mediaType, setETag);
     // Update content-type header according to the priority. (Highest to lowest)
     // 1. MediaType field in response record
     // 2. Payload annotation mediaType value
     // 3. The content type header included in headers field
     // 4. Default content type related to payload
-    string? mediaType = message?.mediaType;
     if (mediaType is string) {
         response.setHeader(CONTENT_TYPE, mediaType);
         return response;
     }
-    if (returnMediaType is string) {
-        response.setHeader(CONTENT_TYPE, returnMediaType);
-    }
     return response;
 }
 
-isolated function setPayload(anydata payload, Response response, boolean setETag = false) {
+isolated function retrieveMediaType(StatusCodeResponse resp, string? retrievedMediaType) returns string? {
+    string? mediaType = resp?.mediaType;
+    if (mediaType is string) {
+        return mediaType;
+    }
+    if (retrievedMediaType is string) {
+        return retrievedMediaType;
+    }
+    return ();
+}
+
+isolated function setPayload(anydata payload, Response response, string? mediaType = (), boolean setETag = false) {
     if payload is () {
         return;
-    } else if payload is xml {
+    }
+
+    if payload is xml {
         response.setXmlPayload(payload);
-        if setETag {
-            response.setETag(payload);
-        }
     } else if payload is string {
         response.setTextPayload(payload);
-        if setETag {
-            response.setETag(payload);
-        }
     } else if payload is byte[] {
         response.setBinaryPayload(payload);
-        if setETag {
-            response.setETag(payload);
-        }
     } else {
-        castToJsonAndSetPayload(response, payload, "anydata to json conversion error: ", setETag);
+        processAnydata(response, payload, mediaType);
+    }
+
+    if setETag {
+        updateETag(response, payload);
     }
 }
 
-isolated function castToJsonAndSetPayload(Response response, anydata payload, string errMsg, boolean setETag = false) {
-    var result = trap val:toJson(payload);
-    if result is error {
-        panic error InitializingOutboundResponseError(errMsg + result.message(), result);
-    } else {
-        response.setJsonPayload(result);
-        if setETag {
-            response.setETag(result);
+isolated function processAnydata(Response response, anydata payload, string? mediaType = ()) {
+    match mediaType {
+        mime:APPLICATION_FORM_URLENCODED => {
+            if payload is map<string> {
+                string result = retrieveUrlEncodedData(payload);
+                response.setTextPayload(result, mime:APPLICATION_FORM_URLENCODED);
+            } else {
+                panic error InitializingOutboundResponseError("unsupported content for application/x-www-form-urlencoded media type");
+            }
```

Comment:
Lets remove the else part

Reply:
Updated the logic as we discussed [a8394959d9c66356c9398959c74c263495ac19f7](https://github.com/ballerina-platform/module-ballerina-http/pull/830/commits/a8394959d9c66356c9398959c74c263495ac19f7)